### PR TITLE
first_hero_wins addon

### PR DESCRIPTION
**Pull request by frankystone**
Not sure if this is ready to merge...

* I am unsure about the name, maybe it is better to name it "first heroes wins" instead of "first hero wins"?
* I am also unsure if i made the localization correct.
* This is also not ready for teams.

But it should be sync_safe, imho.

Any suggestions are welcome :-)

### CODE REVIEW

**Review by Noordfrees — COMMENTED** [review]
A couple of i18n/string fixes/suggestions inline, not tested

**Comment by Noordfrees** [comment]
You also need to specify a `textdomain` key (see `100-wells.wad/init.lua`), and push/pop the textdomain within the script whenever an i18n function is called:
``` diff
+ push_textdomain("first_hero_wins.wad", true)
  scores = scores .. _("Some text")
+ pop_textdomain()
```

File: addons/first_hero_wins.wad/addon [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[global]
+name=_"First Two Hero Wins"
+description=_"The one who has at least 4 heroes in his warehouses wins the game"
+author="kaputtnik"
+version="1.0.0"
+category="win_condition"
+requires=
+min_wl_version=
+max_wl_version=
+sync_safe = true
```
Comment (Noordfrees) [inline addons/first_hero_wins.wad/addon:10]:
```suggestion
sync_safe="true"
```
Widelands profiles are very strict about syntax

File: addons/first_hero_wins.wad/init.lua [hunk 2]
```diff
@@ -0,0 +1,151 @@
+include "scripting/coroutine.lua"
+include "scripting/win_conditions/win_condition_functions.lua"
+include "scripting/win_conditions/win_condition_texts.lua"
+include "scripting/richtext.lua"
+
+local debug = true
```
Comment (Noordfrees) [inline addons/first_hero_wins.wad/init.lua:6]: I guess you'll want to disable this before merging?

**Comment by frankystone** [comment]
Damn, so many suggestions... :-( Sorry that you have so many work with my work...

A question about `_"string"` vs. `_("string")` vs. `_[[string]]`. In this addon i have decided to use `[[ ]]` for all strings with localization and `" "` for strings without localization. The idea was to distinguish between those strings so on proofreading one can see directly which strings have localization.
In our scenarios all strings are done via `[[ ]]` so i thought this is correct. Over there the localization function `_()` misses also the parentheses. See. e.g. [here](https://github.com/widelands/widelands/blob/master/data/campaigns/tutorial01_basic_control.wmf/scripting/texts.lua)

Is it just a matter of taste which one of all the possibilities to use?

**Comment by Noordfrees** [comment]
`"abc"` and `[[abc]]` are effectively equivalent. There's an informal rule that `[[]]` is used for very long texts, such as scenario story windows, and `""` for everything else, but it's not strictly kept.

I generally recommend always using optional `()` around functions for readability, except for pure one-liners such as `include`. Especially if you do further processing on a string (such as calling `bformat`) it is not intuitively apparent whether
```
_"Foo %1%":bformat("bar")
```
evaluates to
```
_("Foo %1%"):bformat("bar")
```
or
```
_("Foo %1%":bformat("bar"))
```
so the parentheses should really not be omitted here. (If I knew enough python I'd write a codecheck rule to enforce this on the official repo as well. Leaving them out is really bad style.)

**Comment by frankystone** [comment]
> You also need to specify a `textdomain` key (see `100-wells.wad/init.lua`), and push/pop the textdomain within the script whenever an i18n function is called:
> 
> ```diff
> + push_textdomain("first_hero_wins.wad", true)
>   scores = scores .. _("Some text")
> + pop_textdomain()
> ```

Hm, `100-wells.wad/init.lua` does contain only one `push/pop_textdomain()` pair and here i have used also one. So i am unsure what you are referring to.

**Comment by Noordfrees** [comment]
I meant two different things:
- The win condition's table needs to contain a `textdomain` key, like in `100-wells`.
- The second issue is something that is not seen in `100-wells` because I don't have translatable strings in the scripting there. But you can see it e.g. in `collectors.lua` line 101-109: all translatable strings need to be surrounded with a pair of textdomain push/pop.

**Comment by frankystone** [comment]
Ah, thanks :-)

Beside your suggestions i will do also some refactoring.

**Comment by frankystone** [comment]
I have made so many changes now and renamed it, so i will propose a new pr soon

**Pull request closed without merging**